### PR TITLE
[BPF] ipv6 dataplane code passes first unittests

### PR DESCRIPTION
**Pull request by tomastigera**
## Description

This PR removes the legacy ipv6 code that was never really used. It is replaced by code that is mostly compiled from the same sources as ipv4 but for 16 byte addresses. Code that needs to be different for v4/6 version is moved to xxx4.h or xxx6.h files.

refs https://github.com/projectcalico/calico/issues/4736

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf-gpl/connect_balancer_v6.c
Comment (tomastigera) [inline felix/bpf-gpl/connect_balancer_v6.c:5]: ctlb is not addressed in this PR, will be addressed eventually

File: felix/bpf-gpl/bpf.h
Comment (fasaxc) [inline felix/bpf-gpl/bpf.h:16]:
```suggestion
/* CALI_BPF_INLINE must be defined before we include any of our headers. They
```

File: felix/bpf-gpl/conntrack.h
Comment (fasaxc) [inline felix/bpf-gpl/conntrack.h:28]: Not sure what this XXX comment is flagging?

File: felix/bpf-gpl/globals.h
Comment (fasaxc) [inline felix/bpf-gpl/globals.h:37]:
```suggestion
 * userspace, but it has the exact same layout as cali_tc_globals in eBPF when
```

File: felix/bpf-gpl/failsafe.h
Comment (fasaxc) [inline felix/bpf-gpl/failsafe.h:40]: XXX?

File: felix/bpf-gpl/icmp6.h
Comment (fasaxc) [inline felix/bpf-gpl/icmp6.h:9]: Worth dropping a TODO?

File: felix/bpf-gpl/ip_addr.h
Comment (fasaxc) [inline felix/bpf-gpl/ip_addr.h:80]: Found this `#else` was a bit lost, maybe add whitespace and a comment ?
```
// IPv4
```

File: felix/bpf-gpl/nat.h
Comment (fasaxc) [inline felix/bpf-gpl/nat.h:98]:
```suggestion
		CALI_DEBUG("bpf_l4_csum_diff(IP): 0x%x\n", csum);
```

File: felix/bpf-gpl/nat.h
Comment (fasaxc) [inline felix/bpf-gpl/nat.h:105]:
```suggestion
		CALI_DEBUG("bpf_l4_csum_diff(IP): 0x%x\n", csum);
```

File: felix/bpf-gpl/nat.h
Comment (fasaxc) [inline felix/bpf-gpl/nat.h:97]: Every time we touch the checksum update logic it seems to break.  Wonder if there's a good way to check this.  the docs are clear as mud.

File: felix/bpf-gpl/tc.c
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:955]: Can you expand these XXX todos a bit so it's easier to know what's missing in future.  Easy for that to get lost without a signpost!

File: felix/bpf-gpl/tc.c
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:1137]: Does this scratch value get zeroed somewhere? Previously we'd zero it; just worried about leakage between packets.

File: felix/bpf-gpl/tc.c
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:1454]: Worth dropping a todo/signpost

File: felix/bpf/state/map.go
Comment (fasaxc) [inline felix/bpf/state/map.go:125]: Removed the size check?

File: felix/bpf-gpl/nat.h
Comment (tomastigera) [reply felix/bpf-gpl/nat.h:97]: we use one of the described case, I added some comments on how we do it.

File: felix/bpf-gpl/tc.c
Comment (tomastigera) [reply felix/bpf-gpl/tc.c:1137]: It is memset to zero on line 1283 before the first use. It is only declared globally because the one branch of switch would have to be otherwise indented. So it is zeroed now only in that branch.

File: felix/bpf-gpl/nat.h
Comment (fasaxc) [inline felix/bpf-gpl/nat.h:93]:
```suggestion
	 * We must use diff because the replace functions cannot calculate a diff for 16
```

File: felix/bpf-gpl/nat.h
Comment (fasaxc) [inline felix/bpf-gpl/nat.h:95]:
```suggestion
	 * minimal impact on v4.
```

File: felix/bpf-gpl/tc.c
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:1558]:
```suggestion
	 * N.B. we could just remember an update the state, however, forwarding
```

**Review by fasaxc — APPROVED** [review]
Few typos in the new comments (thanks!) but I think it's best to get this in and watch how it does in the E2Es.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Pull request merged**
